GLI Dependency

AnastaZIuk opened the pull request:
Status:

- loader is **done**
- writer is **done**

after the changes it needs performing some tests

Crisspl committed:
MTL loader correctly reading file contents

Crisspl committed:
MTL loader ~done

Not generating shaders yet, but i need more info about this

Crisspl committed:
those parentheses should be there

Crisspl committed:
Added MTL loader to IAssetManager

Crisspl committed:
OBJ loader checkpoint commit

Crisspl committed:
OBJ loader almost done

AnastaZIuk committed:
Move itself to the GLIDep branch, add more cmake options

AnastaZIuk committed:
Add some formats, handle loading as an attempt

devshgraphicsprogramming commented on src/irr/asset/CGLIWriter.h:
```diff
@@ -0,0 +1,50 @@
+
+			uint32_t getSupportedFlags() override { return asset::EWF_NONE | asset::EWF_BINARY | asset::EWF_COMPRESSED; }
+
+			uint32_t getForcedFlags() override { return asset::EWF_NONE | asset::EWF_BINARY | asset::EWF_COMPRESSED; }
```
I wouldnt call ktx, dds and kmg "COMPRESSED" (even though they support dumping of BCn data) by compressed we mean something that requires extra CPU effort to get it in a format ready for GPU to use (like LZMA or ZLIB decompression)

AnastaZIuk replied:
Sure

devshgraphicsprogramming commented on src/irr/asset/CGLILoader.h:
```diff
@@ -0,0 +1,53 @@
+			{
+				SContext(const size_t& sizeInBytes) : sourceCodeBuffer(core::make_smart_refctd_ptr<ICPUBuffer>(sizeInBytes)) {}
+				io::IReadFile* file;
+				core::smart_refctd_ptr<ICPUBuffer> sourceCodeBuffer;
```
dont call it `sourceCode` it would make someone thing you're processing JIT script or a shader.

AnastaZIuk replied:
Right, it's old and i removed it

AnastaZIuk committed:
Perform tests with memory dump, make loader work

but I need to get rid of awful memory leak at least

AnastaZIuk committed:
Small fix, have to get rid of Access violation on grab

Crisspl committed:
OBJ loader changes after review

AnastaZIuk committed:
Remove reference to core smart ref pointer

Crisspl committed:
Added pixel pack/unpack to GL state tracking

Crisspl committed:
Some further improvements/fixes

Crisspl committed:
Merge with anastazluk/GLIdep

AnastaZIuk committed:
Prepare an equation to pixel pulling in writing process

Crisspl committed:
Fixed build errors etc

AnastaZIuk committed:
Prepare myself to merge Criss's stuff

AnastaZIuk committed:
Merge branch 'shaders' of https://github.com/Crisspl/IrrlichtBAW into GLIdep

Merge Criss's work

Crisspl committed:
Adjusted asset mutability mechanism and DS creation/reuse in OBJ loader

AnastaZIuk committed:
Make and assure loader and writer with mipmaps and layers

Crisspl committed:
Merge branch 'GLIdep' of github.com:AnastaZIuk/IrrlichtBAW into shaders

AnastaZIuk committed:
Small fixes, it still need to consider cube arrays issue

Crisspl committed:
Some smaller fixes

Crisspl committed:
More fixes

Crisspl committed:
Added builtin BRDF include loader to engine

AnastaZIuk committed:
Fix cubemaps' validation, add swizzle checking on writing

AnastaZIuk committed:
Merge branch 'shaders' of https://github.com/Crisspl/IrrlichtBAW into GLIdep

Sync with Criss

AnastaZIuk committed:
Remove some references according to Criss' advice

Crisspl committed:
Shaders in OBJ loader (not entirely done) + some further fixes and adjustments

Crisspl committed:
A few minor fixes in OBJ shaders

Crisspl committed:
Fixed fuckton of engine bugs

Some left in dirty state, to be fixed in clean way tomorrow

Crisspl committed:
OBJ loader example

it's finished but i didnt manage to finally run it today, sorry, GPU DS needs a way to access its descriptors, will do tomorrow

Crisspl committed:
Working OBJ loader example (06.meshloaders) + fixed more engine bugs

For now running on mori_knob (materials without textures, however textures should work), but target is sponza
Known issues/TODOs:
* ultra weird one: if app us running over some period of time (seems like 5-10s) and after this time i want to close the app, there's a crash regarding access violation during atomic decrement. If app will be closed before those 5-10s, than no crash happens... (????) This behavior is present on debug build, dont know how about release.
* spirv-cross reports offset of PC.extra (see OBJ frag shaders) as 120, will investigate this later but working workaround is to explicitly state layout(offset=124) for PC.extra. Maybe someone could look into this and tell me if i'm calculating std140 offsets wrong??
* in order to get fine-looking sponza, i have to implement bump mapping into shaders (now there's no bump mapping, normal mapping and no reflection mapping/IBL as well)

Crisspl committed:
Fixed PC.extra offset 'bug' and uniforms (GL push consts) setting bug

Crisspl committed:
insertGLSLExtensionsDefines() must go after resolveIncludeDirectives()

Because shaderc preprocessor replaces some extension names with 1 (probably the ones that were already put into core in declared #version of GLSL, im not sure really)
Also fixed another bug in cpu2gpu for meshbuffers

Crisspl committed:
Updated meshloaders example

Crisspl committed:
silly bug fix

Crisspl committed:
Working OBJ sponza example + fixed more bugs

devshgraphicsprogramming commented on include/irr/core/math/floatutil.h:
```diff
@@ -48,39 +48,55 @@ IRR_FORCE_INLINE matrix4SIMD ROUNDING_ERROR<matrix4SIMD>();
 //! Constant for half of PI.
 template<typename T>
-IRR_FORCE_INLINE T HALF_PI()
+IRR_FORCE_INLINE constexpr T HALF_PI()
```
now the derivations with `vectorSIMDf` as the type won't work @Crisspl

devshgraphicsprogramming replied:
fixed in @crisspl 's latest OBJ branch

devshgraphicsprogramming commented on include/irr/core/math/floatutil.h:
```diff
@@ -48,39 +48,55 @@ IRR_FORCE_INLINE matrix4SIMD ROUNDING_ERROR<matrix4SIMD>();
+namespace impl
+{
+    constexpr uint32_t NAN_U32      = 0x7F800000u;
+    constexpr uint32_t INFINITY_U32 = 0x7FFFFFFFu;
```
this is the wrong way around, NAN has all bits set, INF has 0s in the whole mantissa.

devshgraphicsprogramming replied:
@Crisspl @AnastaZIuk

devshgraphicsprogramming replied:
fixed in @Crisspl 's latest OBJ branch

devshgraphicsprogramming commented on include/irr/core/math/floatutil.h:
```diff
@@ -48,39 +48,55 @@ IRR_FORCE_INLINE matrix4SIMD ROUNDING_ERROR<matrix4SIMD>();
+IRR_FORCE_INLINE bool isnan<float>(float val) 
+{ 
+    const float nan_ = core::nan<float>();
+    return reinterpret_cast<uint32_t&>(val)==reinterpret_cast<const uint32_t&>(nan_); 
```
this is not how it works at all !!!!!!!!!!!!!!!

devshgraphicsprogramming replied:
there's a whole host of bit patterns that corresponds to NAN

devshgraphicsprogramming replied:
fixed in @Crisspl 's latest OBJ branch

devshgraphicsprogramming commented on include/irr/video/IGPUObjectFromAssetConverter.h:
```diff
@@ -330,6 +331,7 @@ auto IGPUObjectFromAssetConverter::create(asset::ICPUImage** const _begin, asset
 		{
 			auto tmpBuff = m_driver->createFilledDeviceLocalGPUBufferOnDedMem(cpuimg->getBuffer()->getSize(),cpuimg->getBuffer()->getPointer());
 			m_driver->copyBufferToImage(tmpBuff.get(),gpuimg.get(),count,cpuimg->getRegions().begin());
+            //TODO mipmap generation
```
you can call glGenerateMipmaps here temporarily, for now.

devshgraphicsprogramming replied:
fixed in @Crisspl 's latest OBJ branch

devshgraphicsprogramming commented on source/Irrlicht/COpenGLDriver.cpp:
```diff
@@ -1626,7 +1629,7 @@ void COpenGLDriver::drawMeshBuffer(const IGPUMeshBuffer* mb)
         return;
 
-    found->updateNextState_vertexInput(mb->getVertexBufferBindings(), mb->getIndexBufferBinding().buffer.get(), nullptr, nullptr);
+    found->updateNextState_vertexInput(mb->getVertexBufferBindings(), mb->getIndexBufferBinding().buffer.get(), found->nextState.vertexInputParams.indirectDrawBuf.get(), found->nextState.vertexInputParams.parameterBuf.get());
```
@Crisspl why the sudden change here?

When we draw a meshbuffer we shall draw from non-indirect sources, when we want to draw indirect then we'll never use the meshbuffer concept.

devshgraphicsprogramming commented on src/irr/asset/CGLIWriter.cpp:
```diff
@@ -0,0 +1,327 @@
+
+			case EF_R16_UNORM: return getTranslatedFinalFormat(FORMAT_R16_UNORM_PACK16);		//GL_R16
+			case EF_R16G16_UNORM: return getTranslatedFinalFormat(FORMAT_RG16_UNORM_PACK16);		//GL_RG16
+			case EF_R16G16B16_UNORM: return getTranslatedFinalFormat(FORMAT_RGB16_UNORM_PACK16);		//GL_RGB16
```
formats which don't fit in Power-Of-Two for a texel/block will never be allowed to be texture formats (only vertex formats)

devshgraphicsprogramming replied:
P.S. Except for 24bit RGB basic formats.

devshgraphicsprogramming commented on src/irr/asset/CGLIWriter.cpp:
```diff
@@ -0,0 +1,327 @@
+
+			case EF_R16_SNORM: return getTranslatedFinalFormat(FORMAT_R16_SNORM_PACK16);		//GL_R16_SNORM
+			case EF_R16G16_SNORM: return getTranslatedFinalFormat(FORMAT_RG16_SNORM_PACK16);		//GL_RG16_SNORM
+			case EF_R16G16B16_SNORM: return getTranslatedFinalFormat(FORMAT_RGB16_SNORM_PACK16);		//GL_RGB16_SNORM
```
formats which don't fit in Power-Of-Two for a texel/block will never be allowed to be texture formats (only vertex formats)

devshgraphicsprogramming commented on src/irr/asset/CGLIWriter.cpp:
```diff
@@ -0,0 +1,327 @@
+
+			case EF_R16_UINT: return getTranslatedFinalFormat(FORMAT_R16_UINT_PACK16);				//GL_R16UI
+			case EF_R16G16_UINT: return getTranslatedFinalFormat(FORMAT_RG16_UINT_PACK16);			//GL_RG16UI
+			case EF_R16G16B16_UINT: return getTranslatedFinalFormat(FORMAT_RGB16_UINT_PACK16);			//GL_RGB16UI
```
formats which don't fit in Power-Of-Two for a texel/block will never be allowed to be texture formats (only vertex formats)

devshgraphicsprogramming commented on src/irr/asset/CGLIWriter.cpp:
```diff
@@ -0,0 +1,327 @@
+
+			case EF_R32_UINT: return getTranslatedFinalFormat(FORMAT_R32_UINT_PACK32);				//GL_R32UI
+			case EF_R32G32_UINT: return getTranslatedFinalFormat(FORMAT_RG32_UINT_PACK32);			//GL_RG32UI
+			case EF_R32G32B32_UINT: return getTranslatedFinalFormat(FORMAT_RGB32_UINT_PACK32);			//GL_RGB32UI
```
formats which don't fit in Power-Of-Two for a texel/block will never be allowed to be texture formats (only vertex formats)

devshgraphicsprogramming commented on src/irr/asset/CGLIWriter.cpp:
```diff
@@ -0,0 +1,327 @@
+
+			case EF_R32_SINT: return getTranslatedFinalFormat(FORMAT_R32_SINT_PACK32);				//GL_R32I
+			case EF_R32G32_SINT: return getTranslatedFinalFormat(FORMAT_RG32_SINT_PACK32);			//GL_RG32I
+			case EF_R32G32B32_SINT: return getTranslatedFinalFormat(FORMAT_RGB32_SINT_PACK32);			//GL_RGB32I
```
formats which don't fit in Power-Of-Two for a texel/block will never be allowed to be texture formats (only vertex formats)

devshgraphicsprogramming commented on src/irr/asset/CGLIWriter.cpp:
```diff
@@ -0,0 +1,327 @@
+			// Floating formats
+			case EF_R16_SFLOAT: return getTranslatedFinalFormat(FORMAT_R16_SFLOAT_PACK16);				//GL_R16F
+			case EF_R16G16_SFLOAT: return getTranslatedFinalFormat(FORMAT_RG16_SFLOAT_PACK16);			//GL_RG16F
+			case EF_R16G16B16_SFLOAT: return getTranslatedFinalFormat(FORMAT_RGB16_SFLOAT_PACK16);			//GL_RGB16F
```
formats which don't fit in Power-Of-Two for a texel/block will never be allowed to be texture formats (only vertex formats)

devshgraphicsprogramming commented on src/irr/asset/CGLIWriter.cpp:
```diff
@@ -0,0 +1,327 @@
+
+			case EF_R32_SFLOAT: return getTranslatedFinalFormat(FORMAT_R32_SFLOAT_PACK32);				//GL_R32F
+			case EF_R32G32_SFLOAT: return getTranslatedFinalFormat(FORMAT_RG32_SFLOAT_PACK32);			//GL_RG32F
+			case EF_R32G32B32_SFLOAT: return getTranslatedFinalFormat(FORMAT_RGB32_SFLOAT_PACK32);			//GL_RGB32F
```
formats which don't fit in Power-Of-Two for a texel/block will never be allowed to be texture formats (only vertex formats)

devshgraphicsprogramming commented on src/irr/asset/CGLIWriter.cpp:
```diff
@@ -0,0 +1,327 @@
+
+			case EF_R64_SFLOAT: return getTranslatedFinalFormat(FORMAT_R64_SFLOAT_PACK64);			//GL_R64F
+			case EF_R64G64_SFLOAT: return getTranslatedFinalFormat(FORMAT_RG64_SFLOAT_PACK64);		//GL_RG64F
+			case EF_R64G64B64_SFLOAT: return getTranslatedFinalFormat(FORMAT_RGB64_SFLOAT_PACK64);		//GL_RGB64F
```
formats which don't fit in Power-Of-Two for a texel/block will never be allowed to be texture formats (only vertex formats)

devshgraphicsprogramming commented on src/irr/asset/CGLIWriter.h:
```diff
@@ -0,0 +1,54 @@
+			bool writeGLIFile(io::IWriteFile* file, const asset::ICPUImageView* imageView);
+
+			bool doesItHaveFaces(const IImageView<ICPUImage>::E_TYPE& type);
+			bool doesItHaveLayers(const IImageView<ICPUImage>::E_TYPE& type);
```
change this to inline and static and define it in the header

AnastaZIuk replied:
you mean `IImageAssetHandlerBase`?

devshgraphicsprogramming replied:
no in the GLIWriter header.

AnastaZIuk replied:
okay, but I'm repeating myself in loader then

devshgraphicsprogramming commented on src/irr/asset/CGLIWriter.cpp:
```diff
@@ -0,0 +1,327 @@
+
+			SAssetWriteContext ctx{ _params, _file };
+
+			const asset::ICPUImageView* imageView = IAsset::castDown<ICPUImageView>(_params.rootAsset);
```
bail out on a fail.

devshgraphicsprogramming commented on src/irr/asset/CGLIWriter.cpp:
```diff
@@ -0,0 +1,327 @@
+				os::Printer::log("WRITING GLI: there is a lack of regions!", file->getFileName().c_str(), ELL_INFORMATION);
+				return false;
+			}
+			if (isBlockCompressionFormat(imageInfo.format))
```
really? why?

AnastaZIuk replied:
so do we want to support however?

devshgraphicsprogramming replied:
obviously yes, thats why we had DDS loader in the first place!

AnastaZIuk replied:
to do https://github.com/buildaworldnet/IrrlichtBAW/pull/411#discussion_r367044582

devshgraphicsprogramming commented on src/irr/asset/CGLIWriter.cpp:
```diff
@@ -0,0 +1,327 @@
+					faces = 1;
+
+				return std::make_pair(layers, faces);
+			};
```
this is all wrong, you won't be able to handle (1d,2d) array textures

CUBEMAP behaves as-if its a 6 layer 2D array
CUBEMAP_ARRAY behaves as-if its a 6*N layer 2D array

devshgraphicsprogramming commented on src/irr/asset/CGLIWriter.cpp:
```diff
@@ -0,0 +1,327 @@
+			size_t gliLevels = imageInfo.mipLevels;
+			std::pair<size_t, size_t> layersAndFacesAmount = getFacesAndLayersAmount();
+
+			gli::texture texture(gliTarget, gliFormatAndSwizzles.first, gliExtent3d, layersAndFacesAmount.first, layersAndFacesAmount.second, gliLevels, gli::texture::swizzles_type{ gliFormatAndSwizzles.second[0], gliFormatAndSwizzles.second[1], gliFormatAndSwizzles.second[2], gliFormatAndSwizzles.second[3] });
```
I have an idea.. why don't you use a switch and use the correct derived classes from this diagram?
http://gli.g-truc.net/0.8.0/api/a00022.html

they will all assign to base `gli::texture`

cause right now you're handling texture arrays all wrong and cubemaps.

AnastaZIuk replied:
I will use it in TGA for sure, but it is enough to let it be as it is now I think

devshgraphicsprogramming commented on src/irr/asset/CGLIWriter.cpp:
```diff
@@ -0,0 +1,327 @@
+
+			auto getFullSizeOfLayer = [&](const uint16_t mipLevel) -> uint64_t
+			{
+				auto region = image->getRegions().begin() + mipLevel;
```
that's absolutely not how regions work!!!!!

devshgraphicsprogramming replied:
also if you need the dimensions of a mip-level just use https://github.com/buildaworldnet/IrrlichtBAW/blob/shader_pipeline/include/irr/asset/IImage.h#L403

devshgraphicsprogramming replied:
ok actually its fine, I see what you're doing

devshgraphicsprogramming replied:
its a size of region, not the whole layer.

devshgraphicsprogramming commented on src/irr/asset/CGLIWriter.cpp:
```diff
@@ -0,0 +1,327 @@
+				}	
+			}
+			
+			if (gli::save(texture, file->getFileName().c_str()))
```
does it save ktx, dds, etc. properly?

devshgraphicsprogramming replied:
this won't work, you need to be able to write to virtual files... etc. so find a GLI function or provide one that overrides `fwrite` or whatever its using internally. Can't just go by filename and trust an external filesystem handler.

Maybe there's already something that uses an `ostream` or `istream` ?

AnastaZIuk replied:
I loaded dds, saved and loaded and memory was fine, so ktx left to check

devshgraphicsprogramming replied:
you're not saving to memory or loading from memory, you are passing the filename :man_facepalming:

AnastaZIuk replied:
and passing a texture with all data needed to save an image
it performs saving to file

AnastaZIuk replied:
https://github.com/buildaworldnet/IrrlichtBAW/pull/411#discussion_r367051458

devshgraphicsprogramming replied:
I want you to save to the `IWriteFile` only way to do it is
```cpp
void* someData = ALLOC(SIZE);
gli::save(texture,someData);
file->write(someData,SIZE);
FREE(someData);
```

likewise for reading

AnastaZIuk replied:
well, in that case it's easy, but I'm forced to use `vector<char>` https://github.com/g-truc/gli/blob/master/gli/core/save_dds.inl#L126 (dds example)

devshgraphicsprogramming commented on src/irr/asset/CGLIWriter.cpp:
```diff
@@ -0,0 +1,327 @@
+
+				for (uint16_t layer = 0; layer < imageInfo.arrayLayers; ++layer)
+				{
+					const uint16_t gliLayer = layersFlag ? layer / 6 : 0;
```
a division by 6 is only applicable to cubemaps, all other array textures no division neeeded!!!!

(this is an error in reasoning present in all of your code here)

devshgraphicsprogramming replied:
still wrong

AnastaZIuk replied:
fixed

devshgraphicsprogramming commented on src/irr/asset/CGLIWriter.cpp:
```diff
@@ -0,0 +1,327 @@
+					const uint16_t gliLayer = layersFlag ? layer / 6 : 0;
+					const uint16_t gliFace = facesFlag ? layer % 6 : 0;
+					memcpy(texture.data(gliLayer, gliFace, region->imageSubresource.mipLevel), ptrBeginningOfRegion + (layer * layerSize), layerSize);
+				}	
```
you're doing it all wrong, a region does not contain all the mip levels also layers are counted differently for non-cubemap textures.

Also you can't do it with a memcpy (well you can but at least one `memcpy` per image-row) because you need to honor the region's defined `bufferRowLength` and `bufferImageHeight` to increment the source pointer properly.

AnastaZIuk replied:
My single region contains all layers and faces on certain single mip level

AnastaZIuk replied:
but not sure what about `bufferImageHeight ` you mean in reference to my code

devshgraphicsprogramming replied:
This will not work
```cpp
memcpy(texture.data(gliLayer, gliFace, region->imageSubresource.mipLevel), ptrBeginningOfRegion + (layer * layerSize), layerSize);
```

if your `bufferImageHeight!=height` or `bufferImageWidth!=width`, you can't copy more than a row at once.

AnastaZIuk replied:
Fixed, but I will touch it yet a little tomorrow

devshgraphicsprogramming commented on src/irr/asset/CGLILoader.cpp:
```diff
@@ -0,0 +1,457 @@
+				case gl::INTERNAL_RGBA_PVRTC_4BPPV2: return getTranslatedFinalFormat();			//GL_COMPRESSED_RGBA_PVRTC_4BPPV2_IMG
+				case gl::INTERNAL_RGBA_PVRTC_2BPPV2: return getTranslatedFinalFormat();			//GL_COMPRESSED_RGBA_PVRTC_2BPPV2_IMG
+				case gl::INTERNAL_ATC_RGB: return getTranslatedFinalFormat();						//GL_ATC_RGB_AMD
+				case gl::INTERNAL_ATC_RGBA_EXPLICIT_ALPHA: return getTranslatedFinalFormat();		//GL_ATC_RGBA_EXPLICIT_ALPHA_AMD
```
why are you not solving for the correct format here (and below) @AnastaZIuk ?

AnastaZIuk replied:
we don't have it in `E_FORMAT` as far as I know :D or maybe it is called different in there

devshgraphicsprogramming replied:
all of these formats are actually these ones
https://github.com/buildaworldnet/IrrlichtBAW/blob/master/include/irr/asset/format/EFormat.h#L152

AnastaZIuk replied:
to do

AnastaZIuk replied:
well, I have to add them on writer yet

devshgraphicsprogramming commented on src/irr/asset/CGLILoader.cpp:
```diff
@@ -0,0 +1,457 @@
+			imageInfo.extent.height = texture.extent().y;
+			imageInfo.extent.depth = texture.extent().z;
+			imageInfo.mipLevels = texture.levels();
+			imageInfo.arrayLayers = texture.faces() * texture.layers();
```
if its a cubemap / cubemap array then this is ok.

But overall for other texture types, its `arrayLayers = texture.layers();`

AnastaZIuk replied:
it's good since if there is no cubemap, faces amount is 1

devshgraphicsprogramming commented on src/irr/asset/CGLILoader.cpp:
```diff
@@ -0,0 +1,457 @@
+			auto image = ICPUImage::create(std::move(imageInfo));
+
+			const auto texelBlockByteSize = asset::getTexelOrBlockBytesize(imageInfo.format);
+			auto texelBuffer = core::make_smart_refctd_ptr<ICPUBuffer>(texture.size());
```
does `gli::texture::size()` include all the mipmaps etc?

AnastaZIuk replied:
yes it does, it includes whole size

devshgraphicsprogramming commented on src/irr/asset/CGLILoader.cpp:
```diff
@@ -0,0 +1,457 @@
+
+		bool CGLILoader::isALoadableFileFormat(io::IReadFile* _file) const
+		{
+			return true; // gli provides a function to load files, but we can check files' signature actually if needed
```
check that stuff, otherwise the system will always proceed to load and fail files that are not of GLI loadable type.

AnastaZIuk replied:
to do

devshgraphicsprogramming commented on src/irr/asset/CGLILoader.cpp:
```diff
@@ -0,0 +1,457 @@
+				return {};
+
+			const char* filename = _file->getFileName().c_str();
+			gli::texture texture = gli::load(filename);
```
this won't work, you need to be able to read from virtual files... etc. so find a GLI function or provide one that overrides `fwrite` or whatever its using internally.

Maybe there's already something that uses an `ostream` or `istream` ?

AnastaZIuk replied:
why? I think it will https://github.com/g-truc/ovtsa/blob/master/external/gli/core/load.inl#L12

devshgraphicsprogramming replied:
because you're supposed to read from the `IReadFile` that is passed as argument, you do realize that it doesn't have to be a file in the filesystem, and could be a virtual file in an archive, OR just a pile of data in RAM that mimicks a file?

AnastaZIuk replied:
well I didn't know that, so I will adjust it

devshgraphicsprogramming commented on src/irr/asset/CGLILoader.cpp:
```diff
@@ -0,0 +1,457 @@
+					const uint16_t gliLayer = texture.layers() > 1 ? layer / 6 : 0;
+					const uint16_t gliFace = texture.faces() > 1 ? layer % 6 : 0;
+
+					assignGLIDataToRegion((reinterpret_cast<uint8_t*>(data) + tmpDataSizePerRegionSum + (layer * layerSize)), texture, gliLayer, gliFace, mipLevel, layerSize);
```
wrong equations for layers and faces, will fail on non cubemap textures.

Crisspl committed:
Fixed floatutil issues + isinf()

Crisspl committed:
Made ColorSpaceTest example work

Crisspl committed:
Fixed a silly but kinda big bug in OBJ loader

AnastaZIuk committed:
Merge branch 'shaders' of https://github.com/Crisspl/IrrlichtBAW into GLIdep

Merging fixes

AnastaZIuk committed:
Fix array layers equation, remove some formats

AnastaZIuk committed:
Adjust GLI dependency to IReadFile and IWriteFile

AnastaZIuk committed:
Add validation signature file, add some formats, fix the equation

Crisspl committed:
Fixed IEmulatedDescriptorSet

Crisspl committed:
Added vertex utils builtin include loader

Crisspl committed:
Some minor changes regarding review

Crisspl committed:
asset_utils.h

Crisspl committed:
OBJ,MTL loader: Reworked DS (and layout) creation + actually caching image views

AnastaZIuk committed:
Add some copression formats, move functions and define them in GLI writer/loader's header

AnastaZIuk commented on src/irr/asset/CGLIWriter.cpp:
```diff
@@ -0,0 +1,359 @@
+
+			// Compressed formats
+			case EF_PVRTC1_4BPP_UNORM_BLOCK_IMG: return getTranslatedFinalFormat(FORMAT_RGBA_PVRTC1_8X8_UNORM_BLOCK32);		//GL_COMPRESSED_RGBA_PVRTC_4BPPV1_IMG
+			case EF_PVRTC1_2BPP_UNORM_BLOCK_IMG: return getTranslatedFinalFormat(FORMAT_RGBA_PVRTC1_8X8_UNORM_BLOCK32);		//GL_COMPRESSED_RGBA_PVRTC_2BPPV1_IMG
```
it and above are wrong for sure, but no idea what they are 
I couldn't find anything about those formats

devshgraphicsprogramming replied:
which ones are wrong then?

devshgraphicsprogramming replied:
?

Crisspl committed:
OBJ loader works again, some fixes regarding review

Some common shader parts moved to builtin includes. Next thing TODO: stop dynamically generating frag shaders

Crisspl committed:
OBJ loader now creates just 4 shaders at startup and no more dynamically

Crisspl committed:
A few very minor fixes

devshgraphicsprogramming commented on src/irr/asset/CGLILoader.cpp:
```diff
@@ -0,0 +1,552 @@
+				case gl::INTERNAL_RGBA64F_EXT: return getTranslatedFinalFormat(EF_R64G64B64A64_SFLOAT);		//GL_RGBA64F
+
+				// sRGB formats
+				case gl::INTERNAL_SR8: return getTranslatedFinalFormat(EF_R8_SRGB);				//GL_SR8_EXT
```
This also concerns #412 , if you could add a utility that expands a R8_SRBG texture into a RGB8_SRGB texture to ImageLoaderBase (or whatever that common class with hidden utilities is called) and use this in GLI and TGA loader whenever an R8_SRGB format is encountered.

The reasoning for this is that the R8 format with SRGB encoding used to be available in compatibility (with 2.1) OpenGL profile, but got removed from 3.3 core profile only to be added back to OpenGL 4.6, so not many GPU's drivers can natively accept and display that... so we need a software conversion.

devshgraphicsprogramming replied:
this pretty much needs to happen on any loader that "should" produce a R8_SRGB

devshgraphicsprogramming commented on src/irr/asset/CGLILoader.cpp:
```diff
@@ -0,0 +1,552 @@
+				case gl::INTERNAL_SRGB8_ALPHA8: return getTranslatedFinalFormat(EF_R8G8B8A8_SRGB);		//GL_SRGB8_ALPHA8
+
+				// Packed formats
+				case gl::INTERNAL_RGB9E5: return getTranslatedFinalFormat();			//GL_RGB9_E5
```
https://github.com/buildaworldnet/IrrlichtBAW/blob/master/include/irr/asset/format/EFormat.h#L149

devshgraphicsprogramming commented on src/irr/asset/CGLILoader.cpp:
```diff
@@ -0,0 +1,552 @@
+
+				// Packed formats
+				case gl::INTERNAL_RGB9E5: return getTranslatedFinalFormat();			//GL_RGB9_E5
+				case gl::INTERNAL_RG11B10F: return getTranslatedFinalFormat();			//GL_R11F_G11F_B10F
```
https://github.com/buildaworldnet/IrrlichtBAW/blob/master/include/irr/asset/format/EFormat.h#L148

devshgraphicsprogramming commented on src/irr/asset/CGLILoader.cpp:
```diff
@@ -0,0 +1,552 @@
+				// Packed formats
+				case gl::INTERNAL_RGB9E5: return getTranslatedFinalFormat();			//GL_RGB9_E5
+				case gl::INTERNAL_RG11B10F: return getTranslatedFinalFormat();			//GL_R11F_G11F_B10F
+				case gl::INTERNAL_RG3B2: return getTranslatedFinalFormat();			//GL_R3_G3_B2
```
this we dont support

devshgraphicsprogramming commented on src/irr/asset/CGLILoader.cpp:
```diff
@@ -0,0 +1,552 @@
+				case gl::INTERNAL_RG3B2: return getTranslatedFinalFormat();			//GL_R3_G3_B2
+				case gl::INTERNAL_R5G6B5: return getTranslatedFinalFormat();			//GL_RGB565
+				case gl::INTERNAL_RGB5A1: return getTranslatedFinalFormat();			//GL_RGB5_A1
+				case gl::INTERNAL_RGBA4: return getTranslatedFinalFormat();			//GL_RGBA4
```
rest are here
https://github.com/buildaworldnet/IrrlichtBAW/blob/master/include/irr/asset/format/EFormat.h#L27

devshgraphicsprogramming commented on src/irr/asset/CGLILoader.cpp:
```diff
@@ -0,0 +1,552 @@
+				case gl::INTERNAL_RGB5A1: return getTranslatedFinalFormat();			//GL_RGB5_A1
+				case gl::INTERNAL_RGBA4: return getTranslatedFinalFormat();			//GL_RGBA4
+
+				case gl::INTERNAL_RG4_EXT: return getTranslatedFinalFormat();
```
https://github.com/buildaworldnet/IrrlichtBAW/blob/master/include/irr/asset/format/EFormat.h#L27

devshgraphicsprogramming commented on src/irr/asset/CGLILoader.cpp:
```diff
@@ -0,0 +1,552 @@
+				case gl::INTERNAL_LA8: return getTranslatedFinalFormat();				//GL_LUMINANCE8_ALPHA8
+				case gl::INTERNAL_L16: return getTranslatedFinalFormat();				//GL_LUMINANCE16
+				case gl::INTERNAL_A16: return getTranslatedFinalFormat();				//GL_ALPHA16
+				case gl::INTERNAL_LA16: return getTranslatedFinalFormat();				//GL_LUMINANCE16_ALPHA16
```
these can be expanded to RGB8_SRGB and RGBA8_SRGB

AnastaZIuk replied:
but what about `LA4` then?

AnastaZIuk replied:
there is no enough SRGB `EF_FORMAT`s to cover it

devshgraphicsprogramming replied:
LA4 is 4 bits, just expand the format into 8bits per channel and pretend the format is RGBA8_SRGB.

AnastaZIuk replied:
to fix in next commit in that case

AnastaZIuk replied:
will `EF_R4G4B4A4_UNORM_PACK16` fit it?

devshgraphicsprogramming replied:
not really, because for every two LA4  pixels you'd be getting one `EF_R4G4B4A4_UNORM_PACK16` pixel

devshgraphicsprogramming replied:
so padding needed.

devshgraphicsprogramming replied:
also SRGB concerns

devshgraphicsprogramming commented on src/irr/asset/CGLILoader.cpp:
```diff
@@ -0,0 +1,552 @@
+				case gl::INTERNAL_D32: return getTranslatedFinalFormat();				//GL_DEPTH_COMPONENT32
+				case gl::INTERNAL_D32F: return getTranslatedFinalFormat();				//GL_DEPTH_COMPONENT32F
+				case gl::INTERNAL_D32FS8X24: return getTranslatedFinalFormat();		//GL_DEPTH32F_STENCIL8
+				case gl::INTERNAL_S8_EXT: return getTranslatedFinalFormat();			//GL_STENCIL_INDEX8
```
https://github.com/buildaworldnet/IrrlichtBAW/blob/master/include/irr/asset/format/EFormat.h#L18

AnastaZIuk committed:
Add the utility function, fix equation in the writer and a few docs words

The utility function allows you to convert EF_R8_SRGB to EF_R8G8B8_SRGB
and fills values of G and B channels with R value (RRR)

devshgraphicsprogramming commented on include/irr/asset/IImageLoader.h:
```diff
@@ -19,6 +19,52 @@ namespace irr
+					{
+						for (uint64_t yPos = 0; yPos < imageSize.height; ++yPos)
+							for (uint64_t xPos = 0; xPos < imageSize.width; ++xPos)
+								for (uint64_t zPos = 0; zPos < imageSize.depth; ++zPos)
```
Correct order for such loops is Z,Y,X essentially most major memory stride to most minor.

Read about cache trashing on the internet.

AnastaZIuk replied:
I'm reading, to fix

devshgraphicsprogramming commented on include/irr/asset/IImageLoader.h:
```diff
@@ -19,6 +19,52 @@ namespace irr
+					constexpr auto channelAmount = 3;
+					const void* planarData[] = { redChannelDataLayer, nullptr, nullptr, nullptr };
+					uint8_t* out = nullptr;
+					const size_t texelAmount = getFormatChannelCount(image->getCreationParameters().format);
```
maybe assert that the output image format is EF_R8G8B8_SRGB ?

devshgraphicsprogramming replied:
texelAmount should be the full count of pixels to convert (basically imageSizeWithPitch.x*imageSizeWithPitch.y*imageSizeWithPitch.z)

AnastaZIuk replied:
logic error, I meant channels in texel, but it's wrong

devshgraphicsprogramming commented on include/irr/asset/IImageLoader.h:
```diff
@@ -19,6 +19,52 @@ namespace irr
+					imageSizeWithPitch.Y = region.bufferImageHeight > 0 ? region.bufferImageHeight : region.imageExtent.height;
+					imageSizeWithPitch.Z = region.imageExtent.depth;
+
+					video::convertColor<EF_R8_SRGB, EF_R8G8B8_SRGB>(planarData, out, texelAmount, imageSizeWithPitch);
```
there's no need for any conversion here... R8_SRGB is the same thing as R8G8B8_SRGB, just less channels.

devshgraphicsprogramming replied:
just count your offsets right in `redChannelDataLayer` and you'll be fine (watch out, you should take input pitches/strides as certain formats that rely on Discrete Cosine Transform or block compression will pad rows out [height padding possible too in theory] ).

AnastaZIuk replied:
yeah I know, but I thought that it will be easy to expand the buffer in that way
but I will just get the size and multiple through 3 including all you mentioned

AnastaZIuk replied:
Actually I thought about it as I will get from `convertColor` a ready to use pointer with appropriate size, you're right

AnastaZIuk replied:
have to read about `Discrete Cosine Transform`

devshgraphicsprogramming replied:
well it doesnt matter what DCT is, its just that some image formats will require physical images to be multiples of some sort of a size because of compression or encoding, so padding will be added for the "unnatural" image resolutions.

Example HEVC or h.264 requires block sizes of 8x8 or something like that so you cant have a 400x300 video

devshgraphicsprogramming commented on include/irr/asset/IImageLoader.h:
```diff
@@ -19,6 +19,52 @@ namespace irr
+				{
+					constexpr auto channelAmount = 3;
+					const void* planarData[] = { redChannelDataLayer, nullptr, nullptr, nullptr };
+					uint8_t* out = nullptr;
```
this function will crash, you're not initializing `out` to anything

devshgraphicsprogramming commented on include/irr/asset/IImageLoader.h:
```diff
@@ -19,6 +19,52 @@ namespace irr
+
+				uint8_t* convertR8SRGBdataIntoRGB8SRGBAAndGetIt(const void* redChannelDataLayer, const core::smart_refctd_ptr<ICPUImage>& image, const irr::asset::IImage::SBufferCopy& region)
+				{
+					constexpr auto channelAmount = 3;
```
call it `pixelByteSize`

devshgraphicsprogramming commented on src/irr/asset/CGLILoader.cpp:
```diff
@@ -0,0 +1,552 @@
+				case gl::INTERNAL_S8_EXT: return getTranslatedFinalFormat();			//GL_STENCIL_INDEX8
+
+				// Compressed formats
+				case gl::INTERNAL_RGB_DXT1: return getTranslatedFinalFormat();						//GL_COMPRESSED_RGB_S3TC_DXT1_EXT
```
https://github.com/buildaworldnet/IrrlichtBAW/blob/master/include/irr/asset/format/EFormat.h#L152

AnastaZIuk replied:
but there is a DXT, and our prefix is BC(X). Is it the same?

devshgraphicsprogramming replied:
DXT1-5 and BC1-3 are different names for the same thing, DXT2 and DXT4 are not even formats in their own right.

AnastaZIuk replied:
I will study it

devshgraphicsprogramming commented on src/irr/asset/CGLILoader.cpp:
```diff
@@ -0,0 +1,552 @@
+
+				// Compressed formats
+				case gl::INTERNAL_RGB_DXT1: return getTranslatedFinalFormat();						//GL_COMPRESSED_RGB_S3TC_DXT1_EXT
+				case gl::INTERNAL_RGBA_DXT1: return getTranslatedFinalFormat();					//GL_COMPRESSED_RGBA_S3TC_DXT1_EXT
```
https://github.com/buildaworldnet/IrrlichtBAW/blob/master/include/irr/asset/format/EFormat.h#L154

devshgraphicsprogramming commented on src/irr/asset/CGLILoader.cpp:
```diff
@@ -0,0 +1,552 @@
+				case gl::INTERNAL_RGBA_ASTC_12x12: return getTranslatedFinalFormat(EF_ASTC_12x12_UNORM_BLOCK);				//GL_COMPRESSED_RGBA_ASTC_12x12_KHR
+
+				// sRGB formats
+				case gl::INTERNAL_SRGB_DXT1: return getTranslatedFinalFormat();					//GL_COMPRESSED_SRGB_S3TC_DXT1_EXT
```
https://github.com/buildaworldnet/IrrlichtBAW/blob/master/include/irr/asset/format/EFormat.h#L153

devshgraphicsprogramming commented on src/irr/asset/CGLILoader.cpp:
```diff
@@ -0,0 +1,552 @@
+
+				// sRGB formats
+				case gl::INTERNAL_SRGB_DXT1: return getTranslatedFinalFormat();					//GL_COMPRESSED_SRGB_S3TC_DXT1_EXT
+				case gl::INTERNAL_SRGB_ALPHA_DXT1: return getTranslatedFinalFormat();				//GL_COMPRESSED_SRGB_ALPHA_S3TC_DXT1_EXT
```
https://github.com/buildaworldnet/IrrlichtBAW/blob/master/include/irr/asset/format/EFormat.h#L155

devshgraphicsprogramming commented on src/irr/asset/CGLILoader.cpp:
```diff
@@ -0,0 +1,552 @@
+				// sRGB formats
+				case gl::INTERNAL_SRGB_DXT1: return getTranslatedFinalFormat();					//GL_COMPRESSED_SRGB_S3TC_DXT1_EXT
+				case gl::INTERNAL_SRGB_ALPHA_DXT1: return getTranslatedFinalFormat();				//GL_COMPRESSED_SRGB_ALPHA_S3TC_DXT1_EXT
+				case gl::INTERNAL_SRGB_ALPHA_DXT3: return getTranslatedFinalFormat();				//GL_COMPRESSED_SRGB_ALPHA_S3TC_DXT3_EXT
```
https://github.com/buildaworldnet/IrrlichtBAW/blob/master/include/irr/asset/format/EFormat.h#L157

AnastaZIuk replied:
DXT3 and BC2? Is it valid? I don't get it, what is the convention?

devshgraphicsprogramming replied:
DXT3 is BC2, google it

devshgraphicsprogramming commented on src/irr/asset/CGLILoader.cpp:
```diff
@@ -0,0 +1,552 @@
+				// Compressed formats
+				case gl::INTERNAL_RGB_DXT1: return getTranslatedFinalFormat();						//GL_COMPRESSED_RGB_S3TC_DXT1_EXT
+				case gl::INTERNAL_RGBA_DXT1: return getTranslatedFinalFormat();					//GL_COMPRESSED_RGBA_S3TC_DXT1_EXT
+				case gl::INTERNAL_RGBA_DXT3: return getTranslatedFinalFormat();					//GL_COMPRESSED_RGBA_S3TC_DXT3_EXT
```
https://github.com/buildaworldnet/IrrlichtBAW/blob/master/include/irr/asset/format/EFormat.h#L156

Crisspl committed:
Fixed fix of IEmulatedDescriptorSet

devshgraphicsprogramming commented on src/irr/asset/CGLILoader.cpp:
```diff
@@ -0,0 +1,552 @@
+				case gl::INTERNAL_RGB_DXT1: return getTranslatedFinalFormat();						//GL_COMPRESSED_RGB_S3TC_DXT1_EXT
+				case gl::INTERNAL_RGBA_DXT1: return getTranslatedFinalFormat();					//GL_COMPRESSED_RGBA_S3TC_DXT1_EXT
+				case gl::INTERNAL_RGBA_DXT3: return getTranslatedFinalFormat();					//GL_COMPRESSED_RGBA_S3TC_DXT3_EXT
+				case gl::INTERNAL_RGBA_DXT5: return getTranslatedFinalFormat();					//GL_COMPRESSED_RGBA_S3TC_DXT5_EXT
```
https://github.com/buildaworldnet/IrrlichtBAW/blob/master/include/irr/asset/format/EFormat.h#L158

devshgraphicsprogramming commented on src/irr/asset/CGLILoader.cpp:
```diff
@@ -0,0 +1,552 @@
+				case gl::INTERNAL_SRGB_DXT1: return getTranslatedFinalFormat();					//GL_COMPRESSED_SRGB_S3TC_DXT1_EXT
+				case gl::INTERNAL_SRGB_ALPHA_DXT1: return getTranslatedFinalFormat();				//GL_COMPRESSED_SRGB_ALPHA_S3TC_DXT1_EXT
+				case gl::INTERNAL_SRGB_ALPHA_DXT3: return getTranslatedFinalFormat();				//GL_COMPRESSED_SRGB_ALPHA_S3TC_DXT3_EXT
+				case gl::INTERNAL_SRGB_ALPHA_DXT5: return getTranslatedFinalFormat();				//GL_COMPRESSED_SRGB_ALPHA_S3TC_DXT5_EXT
```
https://github.com/buildaworldnet/IrrlichtBAW/blob/master/include/irr/asset/format/EFormat.h#L159

devshgraphicsprogramming commented on src/irr/asset/CGLILoader.cpp:
```diff
@@ -0,0 +1,552 @@
+				case gl::INTERNAL_RGBA_DXT1: return getTranslatedFinalFormat();					//GL_COMPRESSED_RGBA_S3TC_DXT1_EXT
+				case gl::INTERNAL_RGBA_DXT3: return getTranslatedFinalFormat();					//GL_COMPRESSED_RGBA_S3TC_DXT3_EXT
+				case gl::INTERNAL_RGBA_DXT5: return getTranslatedFinalFormat();					//GL_COMPRESSED_RGBA_S3TC_DXT5_EXT
+				case gl::INTERNAL_R_ATI1N_UNORM: return getTranslatedFinalFormat();				//GL_COMPRESSED_RED_RGTC1
```
https://github.com/buildaworldnet/IrrlichtBAW/blob/master/include/irr/asset/format/EFormat.h#L160

devshgraphicsprogramming commented on src/irr/asset/CGLILoader.cpp:
```diff
@@ -0,0 +1,552 @@
+				case gl::INTERNAL_RGBA_DXT3: return getTranslatedFinalFormat();					//GL_COMPRESSED_RGBA_S3TC_DXT3_EXT
+				case gl::INTERNAL_RGBA_DXT5: return getTranslatedFinalFormat();					//GL_COMPRESSED_RGBA_S3TC_DXT5_EXT
+				case gl::INTERNAL_R_ATI1N_UNORM: return getTranslatedFinalFormat();				//GL_COMPRESSED_RED_RGTC1
+				case gl::INTERNAL_R_ATI1N_SNORM: return getTranslatedFinalFormat();				//GL_COMPRESSED_SIGNED_RED_RGTC1
```
https://github.com/buildaworldnet/IrrlichtBAW/blob/master/include/irr/asset/format/EFormat.h#L161

devshgraphicsprogramming commented on src/irr/asset/CGLILoader.cpp:
```diff
@@ -0,0 +1,552 @@
+				case gl::INTERNAL_RGBA_DXT5: return getTranslatedFinalFormat();					//GL_COMPRESSED_RGBA_S3TC_DXT5_EXT
+				case gl::INTERNAL_R_ATI1N_UNORM: return getTranslatedFinalFormat();				//GL_COMPRESSED_RED_RGTC1
+				case gl::INTERNAL_R_ATI1N_SNORM: return getTranslatedFinalFormat();				//GL_COMPRESSED_SIGNED_RED_RGTC1
+				case gl::INTERNAL_RG_ATI2N_UNORM: return getTranslatedFinalFormat();				//GL_COMPRESSED_RG_RGTC2
```
https://github.com/buildaworldnet/IrrlichtBAW/blob/master/include/irr/asset/format/EFormat.h#L162

devshgraphicsprogramming commented on src/irr/asset/CGLILoader.cpp:
```diff
@@ -0,0 +1,552 @@
+				case gl::INTERNAL_R_ATI1N_UNORM: return getTranslatedFinalFormat();				//GL_COMPRESSED_RED_RGTC1
+				case gl::INTERNAL_R_ATI1N_SNORM: return getTranslatedFinalFormat();				//GL_COMPRESSED_SIGNED_RED_RGTC1
+				case gl::INTERNAL_RG_ATI2N_UNORM: return getTranslatedFinalFormat();				//GL_COMPRESSED_RG_RGTC2
+				case gl::INTERNAL_RG_ATI2N_SNORM: return getTranslatedFinalFormat();				//GL_COMPRESSED_SIGNED_RG_RGTC2
```
https://github.com/buildaworldnet/IrrlichtBAW/blob/master/include/irr/asset/format/EFormat.h#L163

devshgraphicsprogramming commented on src/irr/asset/CGLILoader.cpp:
```diff
@@ -0,0 +1,552 @@
+				case gl::INTERNAL_SRGB8_ALPHA8_ASTC_10x8: return getTranslatedFinalFormat();		//GL_COMPRESSED_SRGB8_ALPHA8_ASTC_10x8_KHR
+				case gl::INTERNAL_SRGB8_ALPHA8_ASTC_10x10: return getTranslatedFinalFormat();		//GL_COMPRESSED_SRGB8_ALPHA8_ASTC_10x10_KHR
+				case gl::INTERNAL_SRGB8_ALPHA8_ASTC_12x10: return getTranslatedFinalFormat();		//GL_COMPRESSED_SRGB8_ALPHA8_ASTC_12x10_KHR
+				case gl::INTERNAL_SRGB8_ALPHA8_ASTC_12x12: return getTranslatedFinalFormat();		//GL_COMPRESSED_SRGB8_ALPHA8_ASTC_12x12_KHR
```
These are all the same as https://github.com/buildaworldnet/IrrlichtBAW/pull/411/files?file-filters%5B%5D=.cpp&file-filters%5B%5D=.h&file-filters%5B%5D=No+extension#diff-28a7c4ae6aba8ec71c707eede79b01e9R477

Except that you swap `UNORM` for `SRGB`

devshgraphicsprogramming commented on src/irr/asset/CGLILoader.cpp:
```diff
@@ -0,0 +1,552 @@
+				case gl::INTERNAL_R_ATI1N_SNORM: return getTranslatedFinalFormat();				//GL_COMPRESSED_SIGNED_RED_RGTC1
+				case gl::INTERNAL_RG_ATI2N_UNORM: return getTranslatedFinalFormat();				//GL_COMPRESSED_RG_RGTC2
+				case gl::INTERNAL_RG_ATI2N_SNORM: return getTranslatedFinalFormat();				//GL_COMPRESSED_SIGNED_RG_RGTC2
+				case gl::INTERNAL_RGB_BP_UNSIGNED_FLOAT: return getTranslatedFinalFormat();		//GL_COMPRESSED_RGB_BPTC_UNSIGNED_FLOAT
```
https://github.com/buildaworldnet/IrrlichtBAW/blob/master/include/irr/asset/format/EFormat.h#L164

devshgraphicsprogramming commented on src/irr/asset/CGLILoader.cpp:
```diff
@@ -0,0 +1,552 @@
+				case gl::INTERNAL_RG_ATI2N_UNORM: return getTranslatedFinalFormat();				//GL_COMPRESSED_RG_RGTC2
+				case gl::INTERNAL_RG_ATI2N_SNORM: return getTranslatedFinalFormat();				//GL_COMPRESSED_SIGNED_RG_RGTC2
+				case gl::INTERNAL_RGB_BP_UNSIGNED_FLOAT: return getTranslatedFinalFormat();		//GL_COMPRESSED_RGB_BPTC_UNSIGNED_FLOAT
+				case gl::INTERNAL_RGB_BP_SIGNED_FLOAT: return getTranslatedFinalFormat();			//GL_COMPRESSED_RGB_BPTC_SIGNED_FLOAT
```
https://github.com/buildaworldnet/IrrlichtBAW/blob/master/include/irr/asset/format/EFormat.h#L165

devshgraphicsprogramming commented on src/irr/asset/CGLILoader.cpp:
```diff
@@ -0,0 +1,552 @@
+				case gl::INTERNAL_SRGB_ALPHA_DXT1: return getTranslatedFinalFormat();				//GL_COMPRESSED_SRGB_ALPHA_S3TC_DXT1_EXT
+				case gl::INTERNAL_SRGB_ALPHA_DXT3: return getTranslatedFinalFormat();				//GL_COMPRESSED_SRGB_ALPHA_S3TC_DXT3_EXT
+				case gl::INTERNAL_SRGB_ALPHA_DXT5: return getTranslatedFinalFormat();				//GL_COMPRESSED_SRGB_ALPHA_S3TC_DXT5_EXT
+				case gl::INTERNAL_SRGB_BP_UNORM: return getTranslatedFinalFormat();				//GL_COMPRESSED_SRGB_ALPHA_BPTC_UNORM
```
https://github.com/buildaworldnet/IrrlichtBAW/blob/master/include/irr/asset/format/EFormat.h#L167

devshgraphicsprogramming commented on src/irr/asset/CGLILoader.cpp:
```diff
@@ -0,0 +1,552 @@
+				case gl::INTERNAL_RG_ATI2N_SNORM: return getTranslatedFinalFormat();				//GL_COMPRESSED_SIGNED_RG_RGTC2
+				case gl::INTERNAL_RGB_BP_UNSIGNED_FLOAT: return getTranslatedFinalFormat();		//GL_COMPRESSED_RGB_BPTC_UNSIGNED_FLOAT
+				case gl::INTERNAL_RGB_BP_SIGNED_FLOAT: return getTranslatedFinalFormat();			//GL_COMPRESSED_RGB_BPTC_SIGNED_FLOAT
+				case gl::INTERNAL_RGB_BP_UNORM: return getTranslatedFinalFormat();					//GL_COMPRESSED_RGBA_BPTC_UNORM
```
https://github.com/buildaworldnet/IrrlichtBAW/blob/master/include/irr/asset/format/EFormat.h#L166

devshgraphicsprogramming commented on src/irr/asset/CGLILoader.cpp:
```diff
@@ -0,0 +1,552 @@
+				case gl::INTERNAL_RGBA_PVRTC_4BPPV1: return getTranslatedFinalFormat(EF_PVRTC1_4BPP_UNORM_BLOCK_IMG);			//GL_COMPRESSED_RGBA_PVRTC_4BPPV1_IMG
+				case gl::INTERNAL_RGBA_PVRTC_2BPPV1: return getTranslatedFinalFormat(EF_PVRTC1_2BPP_UNORM_BLOCK_IMG);			//GL_COMPRESSED_RGBA_PVRTC_2BPPV1_IMG
+				case gl::INTERNAL_RGBA_PVRTC_4BPPV2: return getTranslatedFinalFormat();			//GL_COMPRESSED_RGBA_PVRTC_4BPPV2_IMG
+				case gl::INTERNAL_RGBA_PVRTC_2BPPV2: return getTranslatedFinalFormat();			//GL_COMPRESSED_RGBA_PVRTC_2BPPV2_IMG
```
https://github.com/buildaworldnet/IrrlichtBAW/blob/master/include/irr/asset/format/EFormat.h#L206

all Alpha-containing ones are supported.

devshgraphicsprogramming commented on src/irr/asset/CGLILoader.cpp:
```diff
@@ -0,0 +1,552 @@
+				case gl::INTERNAL_RGB_BP_SIGNED_FLOAT: return getTranslatedFinalFormat();			//GL_COMPRESSED_RGB_BPTC_SIGNED_FLOAT
+				case gl::INTERNAL_RGB_BP_UNORM: return getTranslatedFinalFormat();					//GL_COMPRESSED_RGBA_BPTC_UNORM
+				case gl::INTERNAL_RGB_PVRTC_4BPPV1: return getTranslatedFinalFormat();				//GL_COMPRESSED_RGB_PVRTC_4BPPV1_IMG
+				case gl::INTERNAL_RGB_PVRTC_2BPPV1: return getTranslatedFinalFormat();				//GL_COMPRESSED_RGB_PVRTC_2BPPV1_IMG
```
you can return a special error on those, Vulkan doesn't distinguish between alpha and non alpha containing PVRTC

devshgraphicsprogramming commented on src/irr/asset/CGLILoader.cpp:
```diff
@@ -0,0 +1,552 @@
+				case gl::INTERNAL_SRGB_ALPHA_DXT5: return getTranslatedFinalFormat();				//GL_COMPRESSED_SRGB_ALPHA_S3TC_DXT5_EXT
+				case gl::INTERNAL_SRGB_BP_UNORM: return getTranslatedFinalFormat();				//GL_COMPRESSED_SRGB_ALPHA_BPTC_UNORM
+				case gl::INTERNAL_SRGB_PVRTC_2BPPV1: return getTranslatedFinalFormat();			//GL_COMPRESSED_SRGB_PVRTC_2BPPV1_EXT
+				case gl::INTERNAL_SRGB_PVRTC_4BPPV1: return getTranslatedFinalFormat();			//GL_COMPRESSED_SRGB_PVRTC_4BPPV1_EXT
```
https://github.com/buildaworldnet/IrrlichtBAW/pull/411/files?file-filters%5B%5D=.cpp&file-filters%5B%5D=.h&file-filters%5B%5D=No+extension#r368254851

devshgraphicsprogramming commented on src/irr/asset/CGLILoader.cpp:
```diff
@@ -0,0 +1,552 @@
+				case gl::INTERNAL_SRGB_ALPHA_PVRTC_2BPPV1: return getTranslatedFinalFormat();		//GL_COMPRESSED_SRGB_ALPHA_PVRTC_2BPPV1_EXT
+				case gl::INTERNAL_SRGB_ALPHA_PVRTC_4BPPV1: return getTranslatedFinalFormat();		//GL_COMPRESSED_SRGB_ALPHA_PVRTC_4BPPV1_EXT
+				case gl::INTERNAL_SRGB_ALPHA_PVRTC_2BPPV2: return getTranslatedFinalFormat();		//COMPRESSED_SRGB_ALPHA_PVRTC_2BPPV2_IMG
+				case gl::INTERNAL_SRGB_ALPHA_PVRTC_4BPPV2: return getTranslatedFinalFormat();		//GL_COMPRESSED_SRGB_ALPHA_PVRTC_4BPPV2_IMG
```
https://github.com/buildaworldnet/IrrlichtBAW/pull/411/files?file-filters%5B%5D=.cpp&file-filters%5B%5D=.h&file-filters%5B%5D=No+extension#r368254835

devshgraphicsprogramming commented on src/irr/asset/CGLILoader.cpp:
```diff
@@ -0,0 +1,552 @@
+				case gl::INTERNAL_SRGB_ALPHA_PVRTC_4BPPV2: return getTranslatedFinalFormat();		//GL_COMPRESSED_SRGB_ALPHA_PVRTC_4BPPV2_IMG
+				case gl::INTERNAL_SRGB8_ETC2: return getTranslatedFinalFormat();						//GL_COMPRESSED_SRGB8_ETC2
+				case gl::INTERNAL_SRGB8_PUNCHTHROUGH_ALPHA1_ETC2: return getTranslatedFinalFormat();	//GL_COMPRESSED_SRGB8_PUNCHTHROUGH_ALPHA1_ETC2
+				case gl::INTERNAL_SRGB8_ALPHA8_ETC2_EAC: return getTranslatedFinalFormat();			//GL_COMPRESSED_SRGB8_ALPHA8_ETC2_EAC
```
all here https://github.com/buildaworldnet/IrrlichtBAW/blob/master/include/irr/asset/format/EFormat.h#L196

PUNCHTHROUGH means 1-bit alpha (on or off, like punching through a card)

AnastaZIuk replied:
got it

devshgraphicsprogramming commented on src/irr/asset/CGLILoader.cpp:
```diff
@@ -0,0 +1,552 @@
+				case gl::INTERNAL_RGB16_USCALED_GTC: return getTranslatedFinalFormat();
+				case gl::INTERNAL_RGB16_SSCALED_GTC: return getTranslatedFinalFormat();
+				case gl::INTERNAL_RGBA16_USCALED_GTC: return getTranslatedFinalFormat();
+				case gl::INTERNAL_RGBA16_SSCALED_GTC: return getTranslatedFinalFormat();
```
these look like Vulkan's vertex formats (disallowed for textures) so log an appropriate error message.

devshgraphicsprogramming commented on src/irr/asset/CGLILoader.cpp:
```diff
@@ -0,0 +1,552 @@
+				case gl::INTERNAL_RGB_ETC: return getTranslatedFinalFormat();						//GL_COMPRESSED_RGB8_ETC1
+				case gl::INTERNAL_RGB_ETC2: return getTranslatedFinalFormat(EF_ETC2_R8G8B8_UNORM_BLOCK);						//GL_COMPRESSED_RGB8_ETC2
+				case gl::INTERNAL_RGBA_PUNCHTHROUGH_ETC2: return getTranslatedFinalFormat();		//GL_COMPRESSED_RGB8_PUNCHTHROUGH_ALPHA1_ETC2
+				case gl::INTERNAL_RGBA_ETC2: return getTranslatedFinalFormat(EF_ETC2_R8G8B8A8_UNORM_BLOCK);					//GL_COMPRESSED_RGBA8_ETC2_EAC
```
same as https://github.com/buildaworldnet/IrrlichtBAW/pull/411/files?file-filters%5B%5D=.cpp&file-filters%5B%5D=.h&file-filters%5B%5D=No+extension#r368254941

devshgraphicsprogramming commented on src/irr/asset/CGLILoader.cpp:
```diff
@@ -0,0 +1,552 @@
+				case gl::INTERNAL_ATC_RGBA_EXPLICIT_ALPHA: return getTranslatedFinalFormat();		//GL_ATC_RGBA_EXPLICIT_ALPHA_AMD
+				case gl::INTERNAL_ATC_RGBA_INTERPOLATED_ALPHA: return getTranslatedFinalFormat();	//GL_ATC_RGBA_INTERPOLATED_ALPHA_AMD
+
+				case gl::INTERNAL_RGB_ETC: return getTranslatedFinalFormat();						//GL_COMPRESSED_RGB8_ETC1
```
> • If you want to port a game which is using ETC1 to Vulkan, you can use the existing
resources by using back-word compatible format VK_FORMAT_ETC2_R8G8B8_UNORM_BLOCK

Found that in some Vulkan case study on the web

Crisspl committed:
A little rework of OBJ and MTL loaders

OBJ loader doesnt clone pipelines any more, desc sets are created (and images are loaded) in MTL loader

AnastaZIuk committed:
Avoid cache thrashing, add some formats to the loader

Crisspl committed:
Some minor fixes (mostly sunday review)

Crisspl committed:
stupid replace

Crisspl committed:
Got OBJ loader back to working

Crisspl committed:
more fixes

AnastaZIuk committed:
Add a few formats to the Writer

Crisspl committed:
MTL loader: new shaders, with bsdf_cos_eval() etc...

Crisspl committed:
MTL: large shaders refactor+fixes

TODO:
* better design for calculating alpha and other potential user use-cases
* normalization of blinn-phong specular

Crisspl committed:
WTFFFf wrong shift directions fix

Crisspl committed:
this is not needed for a long time now

AnastaZIuk committed:
Merge branch 'shaders' of https://github.com/Crisspl/IrrlichtBAW into GLIdep

For fake PR

AnastaZIuk committed:
Prepare to change texture iteration assuming blocks

AnastaZIuk committed:
Take pixelOrBlockSize into account

AnastaZIuk committed:
Perform tests on textures using faces, layers and mipmaps

Everything works for non compressed formats. Now I need to port my loader/writer
to use BC formats as well, because now it crashes (I've used DXT5). It seems to me
I need to do something else with size of texture returned from GLI library and also
change a little bit an equation for copying rows to use blocks.

AnastaZIuk committed:
Test it and make it work with BC formats

devshgraphicsprogramming commented on src/irr/asset/CGLIWriter.cpp:
```diff
@@ -0,0 +1,429 @@
+					const auto layerData = texture.data(gliLayer, gliFace, region->imageSubresource.mipLevel);
+					const auto sourceData = ptrBeginningOfRegion + (layer * layerByteSize);
+
+					memcpy
```
aaand its even more broken now!

AnastaZIuk committed:
Nearly done, need setting destination offset of data

devshgraphicsprogramming commented on src/irr/asset/CGLILoader.cpp:
```diff
@@ -124,7 +124,7 @@ namespace irr
 			imageInfo.extent.depth = texture.extent().z;
 			imageInfo.mipLevels = texture.levels();
-			imageInfo.arrayLayers = texture.faces() * texture.layers();
+			imageInfo.arrayLayers = texture.faces() * texture.layers() * imageInfo.mipLevels;
```
WTF are you doing?

devshgraphicsprogramming commented on src/irr/asset/CGLILoader.cpp:
```diff
@@ -148,8 +148,11 @@ namespace irr
+					//region->getDstOffset().x = ... etc
 					region->imageSubresource.mipLevel = regionIndex;
-					region->imageSubresource.layerCount = imageInfo.arrayLayers;
+					region->imageSubresource.layerCount = imageInfo.arrayLayers / imageInfo.mipLevels;
```
WHAT?

devshgraphicsprogramming commented on src/irr/asset/CGLIWriter.cpp:
```diff
@@ -151,16 +145,29 @@ namespace irr
+
+			const auto getTexelBlockDepth = [&](const uint32_t texelDepth)
+			{
+				return texelDepth / texelBlockDimension.Z;
```
these functions need to round up 
`(val+den-1u)/den`

devshgraphicsprogramming commented on src/irr/asset/CGLIWriter.cpp:
```diff
@@ -169,6 +176,18 @@ namespace irr
+				const auto irrTexelBlockHeight = getTexelBlockHeight(texelHeight);
+				const auto irrTexelBlockDepth = texelDepth;
+				
 				for (uint16_t layer = 0; layer < imageInfo.arrayLayers; ++layer)
```
iterate over region layers!

devshgraphicsprogramming commented on src/irr/asset/CGLIWriter.cpp:
```diff
@@ -151,16 +145,29 @@ namespace irr
+				return texelDepth / texelBlockDimension.Z;
+			};
+
 			auto getFullSizeOfLayer = [&](const uint16_t mipLevel) -> uint64_t
```
you don't need this lambda

devshgraphicsprogramming commented on src/irr/asset/CGLIWriter.cpp:
```diff
@@ -177,13 +196,60 @@ namespace irr
+							const auto irrOffset = getPtrOffsetToRowInSpecifiedRegionImageExtent(zPos, yPos);
+							const auto gliOffset = getPtrOffsetToGliData(zPos, yPos);
+
+							if (validYPosFlag)
```
you shall not call boolean functions every loop iteration!

AnastaZIuk committed:
Revert CGLILoader

devshgraphicsprogramming committed:
test this, it should work

AnastaZIuk committed:
Fix some leaks, test everything one more time

AnastaZIuk committed:
Make use of getMipSize()

devshgraphicsprogramming commented on src/irr/asset/CGLIWriter.cpp:
```diff
@@ -164,7 +165,8 @@ namespace irr
 				outOffset = getInBlocks(outOffset);
 
-				core::vector3du32_SIMD outDims(region->imageExtent.width, region->imageExtent.height, region->imageExtent.depth);
+				const auto mipSize = image->getMipSize(regionIndex);
```
still wrong

AnastaZIuk committed:
Use subresource miplevels
